File: common/src/main/java/com/google/tsunami/common/data/NetworkEndpointUtils.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.tsunami.common.data;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.net.HostAndPort;
import com.google.common.net.InetAddresses;
import com.google.tsunami.proto.AddressFamily;
import com.google.tsunami.proto.Hostname;
import com.google.tsunami.proto.IpAddress;
import com.google.tsunami.proto.NetworkEndpoint;
import com.google.tsunami.proto.Port;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;

/** Static utility methods pertaining to {@link NetworkEndpoint} proto buffer. */
public final class NetworkEndpointUtils {
  public static final int MAX_PORT_NUMBER = 65535;

  private NetworkEndpointUtils() {}

  public static boolean hasIpAddress(NetworkEndpoint networkEndpoint) {
    return networkEndpoint.getType().equals(NetworkEndpoint.Type.IP)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_PORT)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_HOSTNAME)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_HOSTNAME_PORT);
  }

  public static boolean hasHostname(NetworkEndpoint networkEndpoint) {
    return networkEndpoint.getType().equals(NetworkEndpoint.Type.HOSTNAME)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.HOSTNAME_PORT)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_HOSTNAME)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_HOSTNAME_PORT);
  }

  public static boolean hasPort(NetworkEndpoint networkEndpoint) {
    return networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_PORT)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.HOSTNAME_PORT)
        || networkEndpoint.getType().equals(NetworkEndpoint.Type.IP_HOSTNAME_PORT);
  }

  public static boolean isIpV6Endpoint(NetworkEndpoint networkEndpoint) {
    return hasIpAddress(networkEndpoint)
        && networkEndpoint.getIpAddress().getAddressFamily().equals(AddressFamily.IPV6);
  }

  /**
   * Converts the given {@link NetworkEndpoint} to its uri authority representation.
   *
   * <p>For example:
   *
   * <ul>
   *   <li>ip_v4 = "1.2.3.4" -&gt; uri = "1.2.3.4"
   *   <li>ip_v6 = "3ffe::1" -&gt; uri = "[3ffe::1]"
   *   <li>host = "localhost" -&gt; url = "localhost"
   *   <li>ip_v4 = "1.2.3.4", port = 8888 -&gt; uri = "1.2.3.4:8888"
   *   <li>ip_v6 = "3ffe::1", port = 8888 -&gt; uri = "[3ffe::1]:8888"
   *   <li>host = "localhost", port = 8888 -&gt; url = "localhost:8888"
   * </ul>
   */
  public static String toUriAuthority(NetworkEndpoint networkEndpoint) {
    return toHostAndPort(networkEndpoint).toString();
  }

  public static HostAndPort toHostAndPort(NetworkEndpoint networkEndpoint) {
    switch (networkEndpoint.getType()) {
      case IP:
        return HostAndPort.fromHost(networkEndpoint.getIpAddress().getAddress());
      case IP_PORT:
        return HostAndPort.fromParts(
            networkEndpoint.getIpAddress().getAddress(), networkEndpoint.getPort().getPortNumber());
      case HOSTNAME:
      case IP_HOSTNAME:
        return HostAndPort.fromHost(networkEndpoint.getHostname().getName());
      case HOSTNAME_PORT:
      case IP_HOSTNAME_PORT:
        return HostAndPort.fromParts(
            networkEndpoint.getHostname().getName(), networkEndpoint.getPort().getPortNumber());
      case UNRECOGNIZED:
      case TYPE_UNSPECIFIED:
        throw new AssertionError("Type for NetworkEndpoint must be specified.");
    }

    throw new AssertionError(
        String.format(
            "Should never happen. Unchecked NetworkEndpoint type: %s", networkEndpoint.getType()));
  }

  /** Returns a {@link NetworkEndpoint} proto buffer object from the given ip address. */
  public static NetworkEndpoint forIp(String ipAddress) {
    checkArgument(InetAddresses.isInetAddress(ipAddress), "'%s' is not an IP address.", ipAddress);

    return NetworkEndpoint.newBuilder()
        .setType(NetworkEndpoint.Type.IP)
        .setIpAddress(
            IpAddress.newBuilder()
                .setAddressFamily(ipAddressFamily(ipAddress))
                .setAddress(ipAddress))
        .build();
  }

  /** Returns a {@link NetworkEndpoint} proto buffer object from the given ip address and port. */
  public static NetworkEndpoint forIpAndPort(String ipAddress, int port) {
    checkArgument(InetAddresses.isInetAddress(ipAddress), "'%s' is not an IP address.", ipAddress);
    checkArgument(
        0 <= port && port <= MAX_PORT_NUMBER,
        "Port out of range. Expected [0, %s], actual %s.",
        MAX_PORT_NUMBER,
        port);

    return forIp(ipAddress).toBuilder()
        .setType(NetworkEndpoint.Type.IP_PORT)
        .setPort(Port.newBuilder().setPortNumber(port))
        .build();
  }

  /** Returns a {@link NetworkEndpoint} proto buffer object from the given hostname. */
  public static NetworkEndpoint forHostname(String hostname) {
    checkArgument(
        !InetAddresses.isInetAddress(hostname), "Expected hostname, got IP address '%s'", hostname);

    return NetworkEndpoint.newBuilder()
        .setType(NetworkEndpoint.Type.HOSTNAME)
        .setHostname(Hostname.newBuilder().setName(hostname))
        .build();
  }

  /**
   * Returns a {@link NetworkEndpoint} proto buffer object from the given ip address and hostname.
   */
  public static NetworkEndpoint forIpAndHostname(String ipAddress, String hostname) {
    return forIp(ipAddress).toBuilder()
        .setType(NetworkEndpoint.Type.IP_HOSTNAME)
        .setHostname(Hostname.newBuilder().setName(hostname))
        .build();
  }

  /** Returns a {@link NetworkEndpoint} proto buffer object from the given hostname and port. */
  public static NetworkEndpoint forHostnameAndPort(String hostname, int port) {
    checkArgument(
        0 <= port && port <= MAX_PORT_NUMBER,
        "Port out of range. Expected [0, %s], actual %s.",
        MAX_PORT_NUMBER,
        port);

    return forHostname(hostname).toBuilder()
        .setType(NetworkEndpoint.Type.HOSTNAME_PORT)
        .setPort(Port.newBuilder().setPortNumber(port))
        .build();
  }

  /**
   * Returns a {@link NetworkEndpoint} proto buffer object from the given ip address, hostname and
   * port.
   */
  public static NetworkEndpoint forIpHostnameAndPort(String ipAddress, String hostname, int port) {
    checkArgument(
        0 <= port && port <= MAX_PORT_NUMBER,
        "Port out of range. Expected [0, %s], actual %s.",
        MAX_PORT_NUMBER,
        port);

    return forIpAndHostname(ipAddress, hostname).toBuilder()
        .setType(NetworkEndpoint.Type.IP_HOSTNAME_PORT)
        .setPort(Port.newBuilder().setPortNumber(port))
        .build();
  }

  /**
   * Returns a {@link NetworkEndpoint} proto buffer object from the given {@code networkEndpoint}
   * and port. The {@code networkEndpoint} parameter cannot contain any port information, otherwise
   * {@link IllegalArgumentException} is thrown.
   */
  public static NetworkEndpoint forNetworkEndpointAndPort(
      NetworkEndpoint networkEndpoint, int port) {
    checkNotNull(networkEndpoint);
    checkArgument(
        0 <= port && port <= MAX_PORT_NUMBER,
        "Port out of range. Expected [0, %s], actual %s.",
        MAX_PORT_NUMBER,
        port);

    switch (networkEndpoint.getType()) {
      case IP:
        return networkEndpoint.toBuilder()
            .setType(NetworkEndpoint.Type.IP_PORT)
            .setPort(Port.newBuilder().setPortNumber(port))
            .build();
      case HOSTNAME:
        return networkEndpoint.toBuilder()
            .setType(NetworkEndpoint.Type.HOSTNAME_PORT)
            .setPort(Port.newBuilder().setPortNumber(port))
            .build();
      case IP_HOSTNAME:
        return networkEndpoint.toBuilder()
            .setType(NetworkEndpoint.Type.IP_HOSTNAME_PORT)
            .setPort(Port.newBuilder().setPortNumber(port))
            .build();
      case IP_PORT:
      case HOSTNAME_PORT:
      case IP_HOSTNAME_PORT:
      case UNRECOGNIZED:
      case TYPE_UNSPECIFIED:
        throw new IllegalArgumentException("Invalid NetworkEndpoint type.");
    }
    throw new AssertionError(
        String.format(
            "Should never happen. Unchecked NetworkEndpoint type: %s", networkEndpoint.getType()));
  }

  public static AddressFamily ipAddressFamily(String ipAddress) {
    InetAddress inetAddress = InetAddresses.forString(ipAddress);

    if (inetAddress instanceof Inet4Address) {
      return AddressFamily.IPV4;
    } else if (inetAddress instanceof Inet6Address) {
      return AddressFamily.IPV6;
    } else {
      throw new AssertionError(String.format("Unknown IP address family for IP '%s'", ipAddress));
    }
  }
}

File: plugin_server/py/common/data/network_endpoint_utils.py
"""Static utility methods pertaining to network endpoint protocol buffer.

For any utility update, please consider if Java's network endpoint utils
(common/src/main/java/com/google/tsunami/common/data/NetworkEndpointUtils.java)
also needs the modification.
"""

import ipaddress
from typing import Optional
from tsunami.proto import network_pb2

AddressFamily = network_pb2.AddressFamily
Hostname = network_pb2.Hostname
IpAddress = network_pb2.IpAddress
NetworkEndpoint = network_pb2.NetworkEndpoint
Type = NetworkEndpoint.Type
DEFINITELY_IP_TYPES = [
    NetworkEndpoint.Type.IP,
    NetworkEndpoint.Type.IP_PORT,
    NetworkEndpoint.Type.IP_HOSTNAME,
    NetworkEndpoint.Type.IP_HOSTNAME_PORT
]
DEFINITELY_HOSTNAME_TYPES = [
    NetworkEndpoint.Type.HOSTNAME,
    NetworkEndpoint.Type.HOSTNAME_PORT,
    NetworkEndpoint.Type.IP_HOSTNAME,
    NetworkEndpoint.Type.IP_HOSTNAME_PORT
]
DEFINITELY_PORT_TYPES = [
    NetworkEndpoint.Type.IP_PORT,
    NetworkEndpoint.Type.IP_HOSTNAME_PORT,
    NetworkEndpoint.Type.HOSTNAME_PORT
]

MAX_PORT_NUMBER = 65535


def has_ip_address(network_endpoint: NetworkEndpoint) -> bool:
  return network_endpoint.type in DEFINITELY_IP_TYPES


def has_hostname(network_endpoint: NetworkEndpoint) -> bool:
  return network_endpoint.type in DEFINITELY_HOSTNAME_TYPES


def has_port(network_endpoint: NetworkEndpoint) -> bool:
  return network_endpoint.type in DEFINITELY_PORT_TYPES


def is_ipv6_endpoint(network_endpoint: NetworkEndpoint) -> bool:
  return has_ip_address(
      network_endpoint
  ) and network_endpoint.ip_address.address_family == AddressFamily.IPV6


def to_uri_authority(network_endpoint: NetworkEndpoint) -> str:
  """Converts network endpoint to URI string.

  Composes URI from the given endpoint information.
  Hostname takes precedence over IP address as hostname is a stable identifier
  for a physical thing whereas IP addresses are ephemeral and unstable.

  Args:
    network_endpoint: instance of a network endpoint protobuf.

  Returns:
    Return None with an unspecified endpoint type. Return URI string with
    valid endpoint type. For example, various combination of ip address, port,
    and hostname would generate the below uri:
      *  ip_v4 = "1.2.3.4" -> uri = "1.2.3.4"
      *  ip_v6 = "3ffe::1" -> uri = "[3ffe::1]"
      *   host = "localhost" -> uri = "localhost"
      *  ip_v4 = "1.2.3.4"   port = 8888 -> uri = "1.2.3.4:8888"
      *  ip_v6 = "3ffe::1"   port = 8888 -> uri = "[3ffe::1]:8888"
      *   host = "localhost" port = 8888 -> uri = "localhost:8888"

  Raises:
    ValueError: an error occurred while looking up network endpoint type.
  """

  ip_address = network_endpoint.ip_address.address
  port = network_endpoint.port.port_number
  hostname = network_endpoint.hostname.name
  uri = ''

  if network_endpoint.type == NetworkEndpoint.Type.TYPE_UNSPECIFIED:
    raise_invalid_network_endpoint_type(network_endpoint.type)
  if network_endpoint.type == NetworkEndpoint.Type.IP:
    uri = ip_to_uri(ip_address)
  elif network_endpoint.type == NetworkEndpoint.Type.IP_PORT:
    uri = ip_to_uri(ip_address) + ':' + str(port)
  elif network_endpoint.type == NetworkEndpoint.Type.IP_HOSTNAME or network_endpoint.type == NetworkEndpoint.Type.HOSTNAME:
    uri = hostname
  elif network_endpoint.type == NetworkEndpoint.Type.IP_HOSTNAME_PORT or network_endpoint.type == NetworkEndpoint.Type.HOSTNAME_PORT:
    uri = hostname + ':' + str(port)
  return uri


def for_ip(ip_address: str) -> NetworkEndpoint:
  """Convert ip address to network endpoint protobuf."""
  network_endpoint = create_with_ip(ip_address)
  network_endpoint.type = NetworkEndpoint.Type.IP
  return network_endpoint


def for_ip_and_port(ip_address: str, port: int) -> NetworkEndpoint:
  """Convert ip address and port to network endpoint protobuf."""
  validate_port(port)
  network_endpoint = create_with_ip(ip_address)
  network_endpoint.type = NetworkEndpoint.Type.IP_PORT
  network_endpoint.port.port_number = port
  return network_endpoint


def for_hostname(hostname: str) -> NetworkEndpoint:
  """Convert hostname to network endpoint protobuf."""
  validate_hostname(hostname)
  network_endpoint = NetworkEndpoint()
  network_endpoint.hostname.name = hostname
  network_endpoint.type = NetworkEndpoint.Type.HOSTNAME
  return network_endpoint


def for_ip_and_hostname(ip_address: str, hostname: str) -> NetworkEndpoint:
  """Convert ip address and hostname to network endpoint protobuf."""
  network_endpoint = create_with_ip(ip_address)
  network_endpoint.type = NetworkEndpoint.Type.IP_HOSTNAME
  network_endpoint.hostname.name = hostname
  return network_endpoint


def for_hostname_and_port(hostname: str, port: int) -> NetworkEndpoint:
  """Convert hostname and port to network endpoint protobuf."""
  validate_port(port)
  validate_hostname(hostname)
  network_endpoint = NetworkEndpoint()
  network_endpoint.hostname.name = hostname
  network_endpoint.type = NetworkEndpoint.Type.HOSTNAME_PORT
  network_endpoint.port.port_number = port
  return network_endpoint


def for_ip_hostname_and_port(ip_address: str, hostname: str,
                             port: int) -> NetworkEndpoint:
  """Convert ip address, hostname and port to network endpoint protobuf."""
  validate_port(port)
  network_endpoint = create_with_ip(ip_address)
  network_endpoint.type = NetworkEndpoint.Type.IP_HOSTNAME_PORT
  network_endpoint.hostname.name = hostname
  network_endpoint.port.port_number = port
  return network_endpoint


def for_network_endpoint_and_port(network_endpoint: NetworkEndpoint,
                                  port: int) -> Optional[NetworkEndpoint]:
  """Create protobuf from endpoint type lookup."""
  validate_port(port)

  if network_endpoint.type == NetworkEndpoint.Type.IP:
    return for_ip_and_port(network_endpoint.ip_address.address, port)
  elif network_endpoint.type == NetworkEndpoint.Type.HOSTNAME:
    return for_hostname_and_port(network_endpoint.hostname.name, port)
  elif network_endpoint.type == NetworkEndpoint.Type.IP_HOSTNAME:
    return for_ip_hostname_and_port(network_endpoint.ip_address.address,
                                    network_endpoint.hostname.name, port)
  else:
    raise_invalid_network_endpoint_type(network_endpoint.type)


def create_with_ip(ip_address: str) -> NetworkEndpoint:
  """Converts ip address to protobuf network endpoint."""
  try:
    ipaddress.ip_address(ip_address)
  except ValueError as exc:
    raise ValueError('%s is not an IP address.' % ip_address) from exc
  return NetworkEndpoint(
      ip_address=IpAddress(
          address=ip_address,
          address_family=address_family(ip_address)))


def validate_port(port: int) -> None:
  if (port < 0 or port > MAX_PORT_NUMBER):
    raise ValueError('Port out of range. Expected [0, %s].' % MAX_PORT_NUMBER)


def validate_hostname(hostname: str) -> None:
  try:
    ipaddress.ip_address(hostname)
    raise Exception("Expected hostname, got IP address '%s'." % hostname)
  except ValueError:
    pass


def raise_invalid_network_endpoint_type(endpoint_type: Type) -> None:
  raise ValueError('Invalid network endpoint type: %s.' %
                   NetworkEndpoint.Type.Name(endpoint_type))


def address_family(ip_address: str) -> AddressFamily:
  try:
    ipaddress.IPv4Address(ip_address)
    return AddressFamily.IPV4
  except ipaddress.AddressValueError:
    return AddressFamily.IPV6


def ip_to_uri(ip_address: str) -> str:
  return ip_address if address_family(
      ip_address) == AddressFamily.IPV4 else '[%s]' % ip_address
